### ISSUE TITLE
resolv.conf should be left untouched when 'Use custom DNS server' is disabled

### ISSUE THREAD

**Issue by sipaktli** [mullvad/mullvadvpn-app#6043]
### Is it a bug?

- [X] I know this is an issue with the app, and contacting Mullvad support is not relevant.

### I have checked if others have reported this already

- [X] I have checked the issue tracker to see if others have reported similar issues.

### Current Behavior

Connection fails if /etc/resolv.conf cannot be modified by mulvad.

### Expected Behavior

If 'Use custom DNS server' is disabled, the application has no reason at all to touch resolv.conf, even if we can use the app to set the same servers or similar configuration.

### Steps to Reproduce

1. chattr +i /etc/resolv.conf
2. Try to connect.

### Additional Information

There are several applications that for whatever reason feel the urge to modify resolv.conf, to the point that some of them even include a nasty 'chattr -i' in their start scripts. For those of us who need to use a custom resolv.conf, it is a really nasty problem when we have to un-chattr the file just to be able to use the vpn. I have had to create a script to permanently monitor and change the file back to what I need.

There is no reason for the modification of this file if the setting to use custom DNS is disabled.

**Comment by ITwrx** [comment]
Exactly. Why is the mullvad vpn client "blocking the internet" when i don't have any content blocking enabled in the Mullvad client? I'm trying to use my own edge firewall's dns (with it's blocking rules) and mullvad is modifying my dns config on the client machines without asking, when i don't want it to, like i'm a ward of the mullvad state. So now, i get BadSite blocking on the client machine from my lan firewall as long as i **don't** use the VPN(!) If i use the vpn, it either overrides my dns and bypasses my firewall rules, or i make my resolv.conf immutable, and the VPN can't "block the internet" and won't even start. Pretty obnoxious.

The "use custom DNS" option only says it will "allow you to add a dns server" when it _needs to be_ a switch for **opting in** to the mullvad dns server, and the related content blocking features. 

I should be able to use the vpn with my own un-touched resolv.conf and local caching dns server(edge firewall) with my preferred upstreams. I'm guessing the mullvad client is changing resolv.conf  by default as an assumed privacy improvement and to make it possible to enable the content blocking features for more typical end users that don't have their own edge firewall, but if mullvad is going to have a feature set like this it needs to be respectful and think about the implementation. As in, don't modify people's dns config without them asking you to. If someone _does_ choose to enable the mullvad content blocking feature, everything needs to be worded and explained better, and they need to be told that mullvad will be overwriting their resolv.conf, so that people don't waste time troubleshooting. I also lost time on this, because i didn't know about the content blocking features, and didn't have them enabled, so i thought my firewall rules weren't working, then i thought systemd on the client was interfering.  I disabled systemd-resolvd and set my own resolv.conf and the mystery `nameserver` entry appeared again... That's when i finally found out Mullvad client was modifying my client machine's dns config even though i had not enabled any features that should require that, as far as i am aware. 

I'm pretty happy with Mullvad's practices, the client, and the vpn so far, but this was poorly thought out/implemented and could come across as condescending and reckless, even if not intended that way.  

Please fix asap. Thanks

**Comment by LeonKuhne** [comment]
^^^^ please fix

**Comment by cinderisles** [comment]
I have the same problem where I want full control over /etc/resolv.conf and don't like that Mullvad tries to take over it, so I hastily made a patch that will just do nothing when it tries to write to resolv.conf. There's a .deb in there for whoever wants it

https://github.com/cinderisles/mullvadvpn-app/releases/tag/unstable

**Comment by hulthe** [comment]
> If 'Use custom DNS server' is disabled, the application has no reason at all to touch resolv.conf

The App is designed to protect the user from leaks that can compromise their privacy. DNS leaks are a kind of issue that is extremely non-obvious to non-technical users, therefore the App is designed to take control of DNS on the users system. This is a feature, and while it may not be ideal for you, it protects the majority of our users.

We welcome suggestions on how to improve our DNS integration, but simply removing it is not something we can do.

Some more context: Since Linux is fragmented by nature, we have a few different ways of managing DNS. In order:
- systemd-resolved
- `resolveconf` cli
- Directly manipulating `/etc/resolv.conf`

Perhaps using `systemd-resolved` of `resolvconf` can satisfy your use-case?

**Comment by ITwrx** [comment]
@hulthe  If mullvad wants to prevent dns leaks for the majority of it's (non-technical) customers that's perfectly understandable, but the app needs to be transparent with it's behavior (mention it's modifying dns config/resolv.conf, even if hidden behind a question mark or something), and at the very least, offer an opt-out for mullvad's dns management that allows truly using one's own dns server where they are in charge of preventing their own dns leaks.

**Comment by hulthe** [comment]
@ITwrx Then I suggest that you open a new issue (or reprhase this one) about e.g. adding an _opt-out of DNS_ feature. We try to prioritize feature requests and bug fixes based on user feedback. Having a clear and actionable GitHub issue that users can engage with helps.

**Comment by ITwrx** [comment]
@hulthe i doubt anyone at mullvad is going to care about a rephrased, closed issue (from paying customers, ignored for nearly 2 years) and i don't know what you think i should rephrase. "Use custom DNS" (existing option in app config) should allow me to *use* my own local, LAN, or public DNS, and *use* that dns' blocking rules. The app already has the ability to *add* a custom DNS, it just doesn't work, at least not  how any user would expect it to; from what i can tell. 

There are already issues in the tracker for this. [1](https://github.com/mullvad/mullvadvpn-app/issues/5564) [2](https://github.com/mullvad/mullvadvpn-app/issues/8392) [3](https://github.com/mullvad/mullvadvpn-app/issues/6080) [4](https://github.com/mullvad/mullvadvpn-app/issues/4125) [5](https://github.com/mullvad/mullvadvpn-app/issues/6594)

These seem to all be solvable by making "use custom dns" work as one would expect. mullvad needs to accept that it has technical users too, and make the escape hatches from "Normal Mode" work properly.

The app should let me *use* my own dns. not just *add* one as a fallback option(?). I don't know what it's doing right now...

**Comment by hulthe** [comment]
@ITwrx I was referring to your suggestion to "offer an opt-out for mullvad's dns management". That particular feature request is not reflected in any of the bug reports you linked.

**Comment by ITwrx** [comment]
@hulthe thanks for the clarification. Now, it seems to me that "Use custom DNS" would satisfy my primary requirements if it actually let one *use their own dns*, instead of adding it to the end of the nameserver list, or not using it at all. IOW, if it worked like people would expect something called "Use custom DNS" to work. My initial reaction about nuking the DNS management from orbit was after a frustrating troubleshooting experience. :)
